File: Homework/CommonResources/CommonResources/MCFrameworkIII/Sde.cs
// Sde.cs
// 
// One-factor Stochastic Differential Equatons (SDEs)
//
// (C) Datasim Education BV 2014
//

using System;
using System.Collections.Generic;

public interface ISde
{ // Standard one-factor SDE dX = a(X,t)dt + b(X,t)dW, X(0)given
  //                         dX = mu(X,t)dt + sig(X,t)dW


    double Drift(double x, double t);          // a (mu)
    double Diffusion(double x, double t);      // b (sig)

    // Some extra functions associated with the SDE
    double DriftCorrected(double x, double t, double B);
    double DiffusionDerivative(double x, double t);

    double InitialCondition
    {
        get;
        set;
    }

    double Expiry
    {
        get;
        set;
    }
}

public class GBM : ISde
{ // Simple SDE

    private double mu;   // Drift
    private double vol;  // Constant volatility
    private double d;    // Constant dividend yield
    private double ic;   // Initial condition
    private double exp;  // Expiry

    public GBM(double driftCoefficient, double diffusionCoefficient, double dividendYield, double initialCondition, double expiry)
    {
        mu = driftCoefficient;
        vol = diffusionCoefficient;
        d = dividendYield;
        ic = initialCondition;
        exp = expiry;
    }

    public double Drift(double x, double t) { return (mu - d) * x; }
    public double Diffusion(double x, double t) { return vol * x; }

    public double DriftCorrected(double x, double t, double B)
    {
       return Drift(x, t) - B * Diffusion(x, t) * DiffusionDerivative(x, t);
    }

    public double DiffusionDerivative(double x, double t)
    {
        return vol;
    }

    // Property to set/get initial condition
    public double InitialCondition
    {
        get
        {
            return ic;
        }
        set
        {
            ic = value;
        }
    }

    // Property to set/get time T
    public double Expiry
    {
        get
        {
            return exp;
        }
        set
        {
            exp = value;
        }
    }

}

public class CEV : ISde
{ // CEV model

    private double mu;   // r
    private double vol;  // Constant volatility
    private double d;    // Constant dividend yield
    private double ic;   // Initial condition
    private double exp;  // Expiry
    private double b;    // Beta

    public CEV(double driftCoefficient, double diffusionCoefficient, double dividendYield, double initialCondition,
               double expiry, double beta)
    {
        mu = driftCoefficient;
        b = beta;
        ic = initialCondition;
        vol = diffusionCoefficient * Math.Pow(ic, 1.0 - b);
        d = dividendYield;
        exp = expiry;
    }

    public double Drift(double x, double t) { return (mu - d) * x; }
    public double Diffusion(double x, double t) { return vol * Math.Pow(x, b); }

    public double DriftCorrected(double x, double t, double B)
    {
        //return drift(X, t) - B * diffusion(X, t) * diffusionDerivative(X, t);
        return Drift(x, t) - B * Diffusion(x, t) * DiffusionDerivative(x, t);
    }

    public double DiffusionDerivative(double x, double t)
    {
        if (b > 1.0)
        {
            return vol * b * Math.Pow(x, b - 1.0);
        }
        else
        {
            return vol * b / Math.Pow(x, 1.0 - b);
        }


    }

    public double InitialCondition
    {
        get
        {
            return ic;
        }
        set
        {
            ic = value;
        }
    }

    public double Expiry
    {
        get
        {
            return exp;
        }
        set
        {
            exp = value;
        }
    }

}

File: Homework/Final/MC/MC/SDE.cpp
//SDE.cpp
//implement methods in SDE class

#include"SDE.hpp"
#include<cmath>
//implement GBM
GBM::GBM(OptionData& src) :
	mu(src.R), vol(src.V), d(src.Div), ic(src.IC), exp(src.T) {}

GBM::GBM(double mu_, double vol_, double d_, double ic_, double exp_) : mu(mu_), vol(vol_), d(d_), ic(ic_), exp(exp_) {}

double GBM::Drift(double x, double t) { return (mu - d)*x; }

double GBM::Diffusion(double x, double t) { return vol*x; }

double GBM::DriftCorrected(double x, double t, double B) { return Drift(x, t) - B*Diffusion(x, t)*DiffusionDerivative(x, t); }

double GBM::DiffusionDerivative(double x, double t) { return vol; }

double GBM::InitialCondition() const { return ic; }
void GBM::InitialCondition(double x) { ic = x; }
double GBM::Expiry() const { return exp; }
void GBM::Expiry(double x) { exp = x; }
//end

//implement CustomizedGBM
CustomizedGBM::CustomizedGBM(OptionData& src):
	GBM(src){}

CustomizedGBM::CustomizedGBM(double mu_, double vol_, double d_, double ic_, double exp_):GBM(mu_,vol_,d_,ic_,exp_){}

double CustomizedGBM::Drift(double x, double t) { 
	return (mu - d)*x*t / exp; //a customized function(x,t) to compute drift
}
double CustomizedGBM::Diffusion(double x, double t) { return vol*x*std::sqrt(1.0 - t / exp); }//customized function(x,t) to get diffusion
double CustomizedGBM::DiffusionDerivative(double x, double t) { return vol*std::sqrt(1.0 - t / exp); }
//end

//implement CEV model
CEV::CEV(double mu_, double vol_, double d_, double ic_, double exp_, double b_):mu(mu_), b(b_), ic(ic_), vol(vol_ * std::pow(ic_, 1.0 - b_)),
                                                                                 d(d_), exp(exp_)
{}
CEV::CEV(OptionData& src, double b_) : mu(src.R), vol(src.V*std::pow(src.IC, 1.0 - b_)), d(src.Div), ic(src.IC), exp(src.T), b(b_) {}
double CEV::Drift(double x, double t) { return (mu - d) * x; }
double CEV::Diffusion(double x, double t) { return vol * std::pow(x, b); }
double CEV::DriftCorrected(double x, double t, double B){return Drift(x, t) - B * Diffusion(x, t) * DiffusionDerivative(x, t);}
double CEV::DiffusionDerivative(double x, double t)
{
	if (b > 1.0)
	{
		return vol * b * std::pow(x, b - 1.0);
	}
	else
	{
		return vol * b / std::pow(x, 1.0 - b);
	}
}
double CEV::InitialCondition() const { return ic; }
void CEV::InitialCondition(double x) { ic = x; }
double CEV::Expiry() const { return exp; }
void CEV::Expiry(double x) { exp = x; }